### PR TITLE
GUI-2368: Charts for instance metrics on auto scaling group details page

### PR DESCRIPTION
**Pull request by kamalgill**
Implements https://eucalyptus.atlassian.net/browse/GUI-2368

### CODE REVIEW

File: eucaconsole/static/js/pages/cloudwatch_charts.js [hunk 1]
```diff
@@ -203,11 +203,12 @@ angular.module('CloudWatchCharts', ['EucaConsoleUtils'])
         document.getElementById('monitoring-form').submit();
     }
 
-    function refreshCharts() {
+    function refreshCharts(refreshOptions) {
+        refreshOptions = refreshOptions || null;
```
Comment (dkavanagh) [inline eucaconsole/static/js/pages/cloudwatch_charts.js]: Scope here is really confusing. Can this be more explicit?

File: eucaconsole/constants/scalinggroups.py [hunk 2]
```diff
@@ -28,51 +28,90 @@
 
 
 SCALING_GROUP_EMPTY_DATA_MESSAGE = _('No data available for this scaling group.')
-SCALING_GROUP_MONITORING_CHARTS_LIST = [
+SCALING_GROUP_MONITORING_CHARTS_LIST = (
```
Comment (coyote240) [inline eucaconsole/constants/scalinggroups.py]: Just out of curiosity, why the switch from a list to a tuple?

File: eucaconsole/constants/scalinggroups.py [hunk 3]
```diff
@@ -28,51 +28,90 @@
...
     {
-        'metric': 'GroupTotalInstances', 'unit': 'None', 'statistic': 'Maximum',
+        'metric': 'GroupTotalInstances', 'unit': 'None', 'statistic': 'Maximum', 'namespace': 'AutoScaling',
```
Comment (coyote240) [inline eucaconsole/constants/scalinggroups.py]: Lots of repetition here.  Seems like we'd benefit from a new-style base class inheriting from dict and setting some basic defaults, capturing only the differences.  Same for the list below.  A small amount of metaclass programming and we can have our collections of carts self-generating, thus doing away with the need for these lists.

Same for the new list below.

File: eucaconsole/static/js/pages/cloudwatch_charts.js [hunk 4]
```diff
@@ -242,10 +243,21 @@ angular.module('CloudWatchCharts', ['EucaConsoleUtils'])
...
     function ChartController($scope, $timeout) {
         renderChart($scope);
-        $scope.$on('cloudwatch:refreshCharts', function () {
-            $timeout(function () {
-                renderChart($scope);
-            });
+        $scope.$on('cloudwatch:refreshCharts', function (evt, refreshOptions) {
+            var doRefresh = false;
+            // Conditionally refresh charts based on passed options
+            if (refreshOptions) {
```
Comment (coyote240) [inline eucaconsole/static/js/pages/cloudwatch_charts.js]: If you're checking for refreshOptions already, you can avoid the extra lookup for refreshOptions.namespace.  If refreshOptions is defined, refreshOptions.namespace will at worst be undefined and throw no key error.

File: eucaconsole/constants/scalinggroups.py (hunk 2 above)
Comment (dkavanagh) [inline eucaconsole/constants/scalinggroups.py]: I wondered same, but tuple is immutable, so perhaps it's simply for accuracy.

Comment (kamalgill) [inline eucaconsole/constants/scalinggroups.py]: module-level constants should be immutable when possible.  Learned that the hard way when I had imported the list and used list.extend(another_list) to observe the resulting list grow with each page reload.  See https://github.com/eucalyptus/eucaconsole/commit/223f141fe56f980e5d846b57f430bb555605c3d5 and https://github.com/eucalyptus/eucaconsole/commit/93eb70969429526cb12cd67bbd924083218f7271

File: eucaconsole/constants/scalinggroups.py (hunk 3 above)
Comment (kamalgill) [inline eucaconsole/constants/scalinggroups.py]: @coyote240: Yeah, good suggestion -- will keep that in mind moving forward.  This structure is converted to JSON and passed to an ng-repeat.  A sequence of dicts seemed to be the simplest approach here, and it's a bit easier for @JMoLo to jump in and update the help text, so I'd vote we leave it as is.

File: eucaconsole/static/js/pages/cloudwatch_charts.js (hunk 4 above)
Comment (kamalgill) [inline eucaconsole/static/js/pages/cloudwatch_charts.js]: @coyote240: Good point -- will simplify.

Comment (kamalgill) [inline eucaconsole/static/js/pages/cloudwatch_charts.js]: @coyote240: On second thought, we'll need the `if (refreshOptions)...` check to handle refreshing charts when the namespace isn't passed.  Simplifying by removing that check broke the desired refresh behavior, so going to leave this one as-is.

Comment (coyote240) [inline eucaconsole/static/js/pages/cloudwatch_charts.js]: You don't remove the `if (refreshOptions)`, remove the refreshOptions.namespace before the &&

**Comment by euca-qa-gui-tester** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.qa1.eucalyptus-systems.com//job/user-console-PR-monitor/1429/
Test FAILed.

**Comment by kamalgill** [comment]
Retest this, please

**Comment by euca-qa-gui-tester** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.qa1.eucalyptus-systems.com//job/user-console-PR-monitor/1433/
Test FAILed.

**Comment by aliceh** [comment]
Retest this, please

**Comment by euca-qa-gui-tester** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.qa1.eucalyptus-systems.com//job/user-console-PR-monitor/1436/
Test FAILed.

**Comment by euca-qa-gui-tester** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.qa1.eucalyptus-systems.com//job/user-console-PR-monitor/1442/
Test FAILed.

**Comment by kamalgill** [comment]
Retest this, please

**Comment by euca-qa-gui-tester** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.qa1.eucalyptus-systems.com//job/user-console-PR-monitor/1444/
Test FAILed.

**Comment by kamalgill** [comment]
Retest this, please

**Comment by euca-qa-gui-tester** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.qa1.eucalyptus-systems.com//job/user-console-PR-monitor/1448/
Test FAILed.

**Pull request merged**